File: frontend/src/i18n/languages.ts
/**
 * Supported UI languages and the rules for picking one.
 *
 * Three things are deliberately kept apart across the app:
 *  1. the **UI language** (this module) — a user preference, never a project one;
 *  2. **public crop-library content**, which has real per-language translations
 *     on shared, language-independent records;
 *  3. **user project content** (project/location/bed names, notes, suppliers),
 *     which is shown exactly as typed and is never translated.
 *
 * To add a language: extend `SUPPORTED_LANGUAGES`, add a locale bundle under
 * `src/i18n/locales/<code>/`, register it in `config.ts`, and add the code to
 * `backend/config/languages.py`.
 */

export const SUPPORTED_LANGUAGES = [
  { code: 'de', label: 'Deutsch' },
  { code: 'en', label: 'English' },
] as const;

export type SupportedLanguage = (typeof SUPPORTED_LANGUAGES)[number]['code'];

/** Last resort when nothing else resolves — matches the backend default. */
export const FALLBACK_LANGUAGE: SupportedLanguage = 'en';

/** Stored preference: an explicit language, or "follow the browser". */
export const AUTO_LANGUAGE = 'auto';
export type LanguagePreference = SupportedLanguage | typeof AUTO_LANGUAGE;

export const LANGUAGE_STORAGE_KEY = 'ui.language';

const SUPPORTED_CODES: readonly string[] = SUPPORTED_LANGUAGES.map((entry) => entry.code);

/**
 * Reduce a BCP-47 tag to a supported language code.
 *
 * `de`, `de-AT`, `de-DE`, `de-CH` and `de_DE` all resolve to `de`; `en`,
 * `en-US`, `en-GB` to `en`; anything unsupported to `null` so callers decide
 * the fallback themselves.
 */
export function normalizeLanguageTag(value: string | null | undefined): SupportedLanguage | null {
  const normalized = (value ?? '').trim().toLowerCase().replace(/_/g, '-');
  if (!normalized) {
    return null;
  }
  const base = normalized.split('-')[0];
  return SUPPORTED_CODES.includes(base) ? (base as SupportedLanguage) : null;
}

export function isSupportedLanguage(value: string | null | undefined): value is SupportedLanguage {
  return normalizeLanguageTag(value) === value;
}

/** Validate a stored/served preference value, defaulting to `auto`. */
export function normalizeLanguagePreference(value: string | null | undefined): LanguagePreference {
  const normalized = (value ?? '').trim().toLowerCase();
  if (normalized === AUTO_LANGUAGE) {
    return AUTO_LANGUAGE;
  }
  return normalizeLanguageTag(normalized) ?? AUTO_LANGUAGE;
}

/**
 * First supported language among the browser's preferences.
 *
 * Only ever used as an *initial* value: a deliberate choice by the user is
 * stored and takes precedence on every later visit, so switching to German on
 * an English-configured browser is not undone by the next page load.
 */
export function detectBrowserLanguage(
  navigatorLike: Pick<Navigator, 'languages' | 'language'> | undefined = typeof navigator === 'undefined'
    ? undefined
    : navigator,
): SupportedLanguage {
  const candidates = [
    ...(navigatorLike?.languages ?? []),
    navigatorLike?.language ?? '',
  ];
  for (const candidate of candidates) {
    const resolved = normalizeLanguageTag(candidate);
    if (resolved) {
      return resolved;
    }
  }
  return FALLBACK_LANGUAGE;
}

/** The guest's stored choice, or `null` when they never made one. */
export function readStoredLanguagePreference(): LanguagePreference | null {
  try {
    const stored = window.localStorage.getItem(LANGUAGE_STORAGE_KEY);
    if (stored === null) {
      return null;
    }
    const normalized = (stored ?? '').trim().toLowerCase();
    if (normalized === AUTO_LANGUAGE) {
      return AUTO_LANGUAGE;
    }
    return normalizeLanguageTag(normalized);
  } catch {
    // Private mode / storage disabled: fall back to browser detection rather
    // than breaking the app.
    return null;
  }
}

export function writeStoredLanguagePreference(preference: LanguagePreference | null): void {
  try {
    if (preference === null) {
      window.localStorage.removeItem(LANGUAGE_STORAGE_KEY);
      return;
    }
    window.localStorage.setItem(LANGUAGE_STORAGE_KEY, preference);
  } catch {
    // Storage failures must not block a language switch; the choice still
    // applies for this session.
  }
}

/**
 * Turn a preference into the language to actually render.
 *
 * `auto` (and "no preference stored") means: follow the browser, falling back
 * to English. An explicit preference always wins over the browser.
 */
export function resolveLanguage(
  preference: LanguagePreference | null | undefined,
  browserLanguage: SupportedLanguage = detectBrowserLanguage(),
): SupportedLanguage {
  if (preference && preference !== AUTO_LANGUAGE) {
    return preference;
  }
  return browserLanguage;
}

/**
 * The language to start the app in, before any user data has loaded.
 *
 * Priority: stored choice (guest or last signed-in session) → browser
 * language → English. The signed-in user's server-side preference is applied
 * on top of this once `/auth/me/` resolves — see `useLanguagePreference`.
 */
export function resolveInitialLanguage(): SupportedLanguage {
  return resolveLanguage(readStoredLanguagePreference());
}

/** Human-readable label for a language, in that language. */
export function getLanguageLabel(code: SupportedLanguage): string {
  return SUPPORTED_LANGUAGES.find((entry) => entry.code === code)?.label ?? code;
}

File: backend/config/languages.py
"""Single source of truth for the languages OpenFarmPlanner supports.

Three separate concerns use these helpers and must not be confused:

1. **UI language** — which language the interface is rendered in. Owned by the
   *user* (``accounts.UserProjectSettings.ui_language``), never by a project,
   so two people can view the same project in different languages.
2. **Public crop-library content** — language-dependent editorial text on
   language-independent records (``crops.CropSpeciesTranslation``,
   ``farm.PublicCropTranslation``).
3. **User project content** — project/location/bed names, private notes,
   supplier names, variety names. Never translated, never duplicated per
   language; stored and shown exactly as typed.

Adding a language means: add the code here, add the frontend locale bundle
under ``frontend/src/i18n/locales/<code>/``, and add the code to
``frontend/src/i18n/languages.ts``. Nothing else is language-specific.
"""

from __future__ import annotations

# Ordered by preference for the "any other translation" fallback step.
SUPPORTED_LANGUAGE_CODES: tuple[str, ...] = ('en', 'de')

# Public crop species become shared lookup data once approved. Keep this list
# explicit so adding a UI language does not silently make every old proposal
# unapprovable until its translations are backfilled.
REQUIRED_PUBLIC_CROP_SPECIES_LANGUAGE_CODES: tuple[str, ...] = ('de', 'en')

# Last-resort UI language, and the second step of the content fallback chain.
DEFAULT_LANGUAGE_CODE = 'en'

# Stored UI preference: an explicit code, or 'auto' = follow the browser.
UI_LANGUAGE_AUTO = 'auto'
UI_LANGUAGE_CHOICES = [
    (UI_LANGUAGE_AUTO, 'Automatic (browser language)'),
    ('de', 'Deutsch'),
    ('en', 'English'),
]
UI_LANGUAGE_VALUES = frozenset({UI_LANGUAGE_AUTO, *SUPPORTED_LANGUAGE_CODES})


def normalize_language_tag(value: str | None) -> str:
    """Reduce a BCP-47 tag to a supported code, or '' when unsupported.

    ``de-AT``/``de_DE``/``DE`` all become ``de``; ``en-GB`` becomes ``en``;
    ``fr`` and nonsense become ``''`` so callers can decide the fallback.
    """
    normalized = (value or '').strip().lower().replace('_', '-')
    if not normalized:
        return ''
    base = normalized.split('-', 1)[0]
    return base if base in SUPPORTED_LANGUAGE_CODES else ''


def resolve_language_tag(value: str | None, *, default: str = DEFAULT_LANGUAGE_CODE) -> str:
    """Like :func:`normalize_language_tag`, but never returns an empty string."""
    return normalize_language_tag(value) or default


def parse_accept_language(header: str | None) -> str:
    """First supported language in an ``Accept-Language`` header, or ''.

    Quality values are honoured, so ``fr;q=0.9, de;q=0.8`` resolves to ``de``
    rather than to the header's first (unsupported) entry.
    """
    if not header:
        return ''
    candidates: list[tuple[float, int, str]] = []
    for index, part in enumerate(header.split(',')):
        piece, _, params = part.strip().partition(';')
        code = normalize_language_tag(piece)
        if not code:
            continue
        quality = 1.0
        for param in params.split(';'):
            key, _, raw_value = param.strip().partition('=')
            if key.strip() == 'q':
                try:
                    quality = float(raw_value)
                except ValueError:
                    quality = 0.0
        candidates.append((-quality, index, code))
    if not candidates:
        return ''
    candidates.sort()
    return candidates[0][2]


def normalize_ui_language_preference(value: str | None) -> str:
    """Validate a stored UI preference value ('auto' | 'de' | 'en')."""
    normalized = (value or '').strip().lower()
    if normalized in UI_LANGUAGE_VALUES:
        return normalized
    raise ValueError(f'Unsupported UI language preference: {value!r}')


def resolve_request_language(request) -> str:
    """The content language an API response should prefer for this request.

    Resolution order, matching the frontend's UI-language order so the two
    cannot disagree:

    1. an explicit ``?language=`` query parameter (unsupported values are
       ignored rather than rejected, so a stale client link never 400s),
    2. the signed-in user's stored preference (unless it is ``auto``),
    3. the ``Accept-Language`` header,
    4. English.
    """
    query_params = getattr(request, 'query_params', None)
    if query_params is None:
        query_params = getattr(request, 'GET', {})
    explicit = normalize_language_tag(query_params.get('language'))
    if explicit:
        return explicit

    user = getattr(request, 'user', None)
    if user is not None and getattr(user, 'is_authenticated', False):
        settings_row = getattr(user, 'project_settings', None)
        stored = getattr(settings_row, 'ui_language', '') or ''
        if stored != UI_LANGUAGE_AUTO:
            preference = normalize_language_tag(stored)
            if preference:
                return preference

    header = ''
    headers = getattr(request, 'headers', None)
    if headers is not None:
        header = headers.get('Accept-Language', '')
    return parse_accept_language(header) or DEFAULT_LANGUAGE_CODE


def build_fallback_chain(language_code: str | None) -> tuple[str, ...]:
    """Content fallback order: requested → English → every other language.

    The literal fallback order from the spec, in one place, so the API,
    serializers and admin views cannot drift apart.
    """
    requested = normalize_language_tag(language_code)
    chain: list[str] = []
    for code in (requested, DEFAULT_LANGUAGE_CODE, *SUPPORTED_LANGUAGE_CODES):
        if code and code not in chain:
            chain.append(code)
    return tuple(chain)


def resolve_translation(
    translations: dict[str, str],
    language_code: str | None,
    *,
    fallback: str = '',
) -> tuple[str, str]:
    """Pick the best available text and report which language it came from.

    :param translations: language code → text (blank values count as missing).
    :param language_code: the requested UI language.
    :param fallback: technical replacement text when nothing is available.
    :return: ``(text, language_code_used)``; the used code is ``''`` when the
        ``fallback`` had to be used, which is how callers know to render the
        "only available in another language" hint.
    """
    for code in build_fallback_chain(language_code):
        text = (translations.get(code) or '').strip()
        if text:
            return text, code
    for code, text in translations.items():
        if (text or '').strip():
            return text.strip(), code
    return fallback, ''
